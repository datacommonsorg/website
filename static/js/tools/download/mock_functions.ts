/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* mocked axios calls for Page test for download tool. */

jest.mock("axios");
import axios from "axios";
import { when } from "jest-when";

import { stringifyFn } from "../../utils/axios";

export function axios_mock(): void {
  // Mock all the async axios call.
  axios.get = jest.fn();
  axios.post = jest.fn();

  const rootGroupsData = {
    data: {
      childStatVarGroups: [
        {
          displayName: "Demographics",
          id: "dc/g/Demographics",
          specializedEntity: "Demographics",
          descendentStatVarCount: 2,
        },
        {
          displayName: "Economics",
          id: "dc/g/Economics",
          specializedEntity: "Economics",
          descendentStatVarCount: 2,
        },
      ],
    },
  };

  const demographicsGroupsData = {
    data: {
      childStatVarGroups: [],
      childStatVars: [
        {
          displayName: "Population",
          id: "Count_Person",
          searchName: "Count Of Person",
          hasData: true,
        },
        {
          displayName: "Median Age",
          id: "Median_Age_Person",
          searchName: "Median Age",
          hasData: true,
        },
      ],
    },
  };

  const csvData =
    "placeDcid,placeName,Date:Count_Person,Value:Count_Person,Source:Count_Person\ngeoId/06001,Alameda County,2021,1648556,https://www2.census.go\ngeoId/06002,Alpine County,2021,1235,https://www2.census.gov/";

  // get statvar properties Count_Person
  when(axios.get)
    .calledWith("/api/stats/stats-var-property?dcid=Count_Person")
    .mockResolvedValue({
      data: {
        Count_Person: {
          md: "",
          mprop: "count",
          pt: "Person",
          pvs: {},
          title: "Population",
        },
      },
    });

  // get statvar properties Median_Age_Person
  when(axios.get)
    .calledWith("/api/stats/stats-var-property?dcid=Median_Age_Person")
    .mockResolvedValue({
      data: {
        Median_Age_Person: {
          md: "",
          mprop: "",
          pt: "Person",
          pvs: {},
          title: "Median Age",
        },
      },
    });

  // get place types, geoId/06
  when(axios.get)
    .calledWith("/api/place/type/geoId/06")
    .mockResolvedValue({ data: "State" });

  // get place names, geoId/06
  when(axios.get)
    .calledWith("/api/place/name?dcids=geoId/06")
    .mockResolvedValue({ data: { "geoId/06": "California" } });

  // get root stat var group
  when(axios.post)
    .calledWith("/api/variable-group/info", {
      dcid: "dc/g/Root",
      entities: [],
      numEntitiesExistence: undefined,
    })
    .mockResolvedValue(rootGroupsData);

  // get root stat var group for places in geoId/06
  when(axios.post)
    .calledWith("/api/variable-group/info", {
      dcid: "dc/g/Root",
      entities: ["geoId/06001", "geoId/06002"],
      numEntitiesExistence: undefined,
    })
    .mockResolvedValue(rootGroupsData);

  when(axios.post)
    .calledWith("/api/variable-group/info", {
      dcid: "dc/g/Root",
      entities: ["geoId/06002", "geoId/06001"],
      numEntitiesExistence: undefined,
    })
    .mockResolvedValue(rootGroupsData);

  // get demographics stat var group for places in geoId/06
  when(axios.post)
    .calledWith("/api/variable-group/info", {
      dcid: "dc/g/Demographics",
      entities: ["geoId/06001", "geoId/06002"],
      numEntitiesExistence: undefined,
    })
    .mockResolvedValue(demographicsGroupsData);

  when(axios.post)
    .calledWith("/api/variable-group/info", {
      dcid: "dc/g/Demographics",
      entities: ["geoId/06002", "geoId/06001"],
      numEntitiesExistence: undefined,
    })
    .mockResolvedValue(demographicsGroupsData);

  // stat var info for demographics node
  when(axios.get)
    .calledWith("/api/variable/info", {
      params: {
        dcids: ["Count_Person", "Median_Age_Person"],
      },
      paramsSerializer: stringifyFn,
    })
    .mockResolvedValue({
      data: {
        Count_Person: {
          placeTypeSummary: {
            type1: {
              numPlaces: 0,
              topPlaces: [],
            },
          },
        },
        Median_Age_Person: {
          placeTypeSummary: {
            type1: {
              numPlaces: 0,
              topPlaces: [],
            },
          },
        },
      },
    });

  // get stat var path for Count_Person
  when(axios.get)
    .calledWith("/api/variable/path?dcid=Count_Person")
    .mockResolvedValue({
      data: ["Count_Person", "dc/g/Demographics"],
    });

  // get stat var path for Median_Age_Person
  when(axios.get)
    .calledWith("/api/variable/path?dcid=Median_Age_Person")
    .mockResolvedValue({
      data: ["Median_Age_Person", "dc/g/Demographics"],
    });

  // get places in for counties in geoId/06
  when(axios.get)
    .calledWith("/api/place/places-in?dcid=geoId/06&placeType=County")
    .mockResolvedValue({
      data: {
        "geoId/06": ["geoId/06001", "geoId/06002"],
      },
    });

  // get parent place for geoId/06
  when(axios.get)
    .calledWith("/api/place/parent/geoId/06")
    .mockResolvedValue({
      data: ["country/USA"],
    });

  // get facets within place for Count_Person
  when(axios.get)
    .calledWith("/api/facets/within", {
      params: {
        childType: "County",
        maxDate: "latest",
        minDate: "latest",
        parentPlace: "geoId/06",
        statVars: ["Count_Person"],
      },
      paramsSerializer: stringifyFn,
    })
    .mockResolvedValue({
      data: {
        Count_Person: {
          importName: "Facet1",
        },
      },
    });

  when(axios.get)
    .calledWith("/api/facets/within", {
      params: {
        childType: "County",
        maxDate: "",
        minDate: "2020",
        parentPlace: "geoId/06",
        statVars: ["Count_Person"],
      },
      paramsSerializer: stringifyFn,
    })
    .mockResolvedValue({
      data: {
        Count_Person: {
          importName: "Facet1",
        },
      },
    });

  // get facets within place for Count_Person and Median_Age_Person
  when(axios.get)
    .calledWith("/api/facets/within", {
      params: {
        childType: "County",
        maxDate: "",
        minDate: "2020",
        parentPlace: "geoId/06",
        statVars: ["Count_Person", "Median_Age_Person"],
      },
      paramsSerializer: stringifyFn,
    })
    .mockResolvedValue({
      data: {
        Count_Person: {
          importName: "Facet1",
        },
        Median_Age_Person: {
          importName: "Facet1",
        },
      },
    });

  // get place stats vars for places in geoId/06
  when(axios.post)
    .calledWith("/api/place/stat-vars/union", {
      dcids: ["geoId/06001", "geoId/06002"],
      statVars: ["Count_Person"],
    })
    .mockResolvedValue({
      data: ["Count_Person"],
    });

  when(axios.post)
    .calledWith("/api/place/stat-vars/union", {
      dcids: ["geoId/06002", "geoId/06001"],
      statVars: ["Count_Person"],
    })
    .mockResolvedValue({
      data: ["Count_Person"],
    });

  // get csv
  when(axios.post)
    .calledWith("/api/csv/within", {
      childType: "County",
      facetMap: { Count_Person: "" },
      maxDate: "latest",
      minDate: "latest",
      parentPlace: "geoId/06",
      rowLimit: 7,
      statVars: ["Count_Person"],
    })
    .mockResolvedValue({
      data: csvData,
    });

  when(axios.post)
    .calledWith("/api/csv/within", {
      childType: "County",
      facetMap: { Count_Person: "" },
      maxDate: "latest",
      minDate: "latest",
      parentPlace: "geoId/06",
      statVars: ["Count_Person"],
    })
    .mockResolvedValue({
      data: csvData,
    });

  when(axios.post)
    .calledWith("/api/csv/within", {
      childType: "County",
      facetMap: { Count_Person: "" },
      maxDate: "",
      minDate: "2020",
      parentPlace: "geoId/06",
      rowLimit: 7,
      statVars: ["Count_Person"],
    })
    .mockResolvedValue({
      data: csvData,
    });

  when(axios.post)
    .calledWith("/api/csv/within", {
      childType: "County",
      facetMap: { Count_Person: "" },
      maxDate: "",
      minDate: "2020",
      parentPlace: "geoId/06",
      statVars: ["Count_Person"],
    })
    .mockResolvedValue({
      data: csvData,
    });
}
