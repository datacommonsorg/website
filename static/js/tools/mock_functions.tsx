/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

jest.mock("axios");
import { when } from "jest-when";
import axios from "axios";
import hierarchy from "../../data/hierarchy_top.json";
import hierarchy_complete from "../../data/hierarchy_statsvar.json";
import { drawGroupLineChart } from "../chart/draw";
import * as d3 from "d3";

export function axios_mock(): void {
  // Mock all the async axios call.
  axios.get = jest.fn();

  // get hierachy.json
  when(axios.get)
    .calledWith("../../data/hierarchy_statsvar.json")
    .mockResolvedValue({ data: hierarchy });

  // get statsvar properties Median_Age_Person
  when(axios.get)
    .calledWith("/api/stats/stats-var-property?dcid=Median_Age_Person")
    .mockResolvedValue({
      data: {
        Median_Age_Person: { md: "", mprop: "age", pt: "Person", pvs: {} },
      },
    });

  // get statsVar info of Median_Age_Person and Count_Person
  when(axios.get)
    .calledWith(
      "/api/stats/stats-var-property?dcid=Median_Age_Person&dcid=Count_Person"
    )
    .mockResolvedValue({
      data: {
        Median_Age_Person: { md: "", mprop: "age", pt: "Person", pvs: {} },
        Count_Person: { md: "", mprop: "count", pt: "Person", pvs: {} },
      },
    });

  // get place stats vars, geoId/05
  when(axios.get)
    .calledWith("/api/place/statsvars/geoId/05")
    .mockResolvedValue({ data: ["Count_Person", "Median_Age_Person"] });

  // get place names, geoId/05
  when(axios.get)
    .calledWith("/api/place/name?dcid=geoId/05")
    .mockResolvedValue({ data: { "geoId/05": "Place" } });

  // get data, geoId/05, Median_Age_Person
  when(axios.get)
    .calledWith("/api/stats/Median_Age_Person?&dcid=geoId/05")
    .mockResolvedValue({
      data: {
        "geoId/05": {
          placeDcid: "geoId/05",
          placeName: "Arkansas",
          provenanceDomain: "census.gov",
          data: { "2011": 37.3, "2012": 37.4, "2013": 37.5, "2014": 37.6 },
        },
      },
    });

  // get data, geoId/05,Count_Person
  when(axios.get)
    .calledWith("/api/stats/Count_Person?&dcid=geoId/05")
    .mockResolvedValue({
      data: {
        "geoId/05": {
          placeDcid: "geoId/05",
          placeName: "Arkansas",
          provenanceDomain: "census.gov",
          data: { "1999": 37.3, "2010": 37.4, "2020": 37.5 },
        },
      },
    });
}

export function mock_hierarchy_complete(): void {
  axios.get = jest.fn();
  when(axios.get)
    .calledWith("../../data/hierarchy_statsvar.json")
    .mockResolvedValue({ data: hierarchy_complete });
}

// Mock drawGroupLineChart() as getComputedTextLength can has issue with jest
// and jsdom.
export function drawGroupLineChart_mock(): void {
  (drawGroupLineChart as jest.Mock).mockImplementation(
    (selector: string | HTMLDivElement) => {
      let container: d3.Selection<any, any, any, any>;
      if (typeof selector === "string") {
        container = d3.select("#" + selector);
      } else if (selector instanceof HTMLDivElement) {
        container = d3.select(selector);
      } else {
        return;
      }
      container.selectAll("svg").remove();

      const svg = container
        .append("svg")
        .attr("width", 500)
        .attr("height", 500);
      svg.append("text").text("svg text");
    }
  );
}
