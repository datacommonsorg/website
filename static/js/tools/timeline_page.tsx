/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { Component } from "react";
import {
  getStatsVarInfo,
  getPlaceNames,
  getStatsVar,
  StatsVarNode,
  StatsVarInfo,
  TimelineParams,
  ChartOptions,
} from "./timeline_util";
import { SearchBar } from "./timeline_search";
import { Menu } from "./statsvar_menu";
import { Info } from "./timeline_info";
import { ChartRegion } from "./timeline_chart_region";
import {
  NoopStatsVarFilter,
  TimelineStatsVarFilter,
  StatsVarFilterInterface,
} from "./commons";
import _ from "lodash";

interface PageStateType {
  statsVarNodes: StatsVarNode;
  statsVarInfo: { [key: string]: StatsVarInfo };
  placeIdNames: Record<string, string>; // [(placeId, placeName)]
  statsVarTitle: Record<string, string>;
  statsVarValid: Set<string>;
  chartOptions: ChartOptions;
}

class Page extends Component<Record<string, unknown>, PageStateType> {
  params: TimelineParams;

  constructor(props: Record<string, unknown>) {
    super(props);
    this.handleHashChange = this.handleHashChange.bind(this);
    this.params = new TimelineParams();
    this.params.getParamsFromUrl();
    // set default statsVarTitle as the statsVar dcids
    const statsVarTitle = {};
    for (const statsVar of this.params.getStatsVarDcids()) {
      statsVarTitle[statsVar] = statsVar;
    }
    this.state = {
      placeIdNames: {},
      statsVarValid: new Set(),
      statsVarNodes: _.cloneDeep(this.params.statsVarNodes),
      statsVarInfo: {},
      statsVarTitle: statsVarTitle,
      chartOptions: _.cloneDeep(this.params.chartOptions),
    };
  }

  componentDidMount(): void {
    window.addEventListener("hashchange", this.handleHashChange);
    this.getAllPromises();
  }

  private handleHashChange(): void {
    if (this.params.listenHashChange) {
      // do not update if it's set by calling add/remove place/statsVar
      this.params.getParamsFromUrl();
      if (
        !_.isEqual(this.params.statsVarNodes, this.state.statsVarNodes) ||
        !_.isEqual(
          this.params.placeDcids,
          Object.keys(this.state.placeIdNames)
        ) ||
        !_.isEqual(this.params.chartOptions, this.state.chartOptions)
      ) {
        this.setState({
          statsVarNodes: _.cloneDeep(this.params.statsVarNodes),
          chartOptions: this.params.chartOptions,
        });
        this.getAllPromises();
      }
    }
    this.params.listenHashChange = true;
  }

  private getAllPromises(): void {
    let statsVarInfoPromise = Promise.resolve({});
    if (this.params.getStatsVarDcids().length !== 0) {
      statsVarInfoPromise = getStatsVarInfo(this.params.getStatsVarDcids());
    }
    let placesPromise = Promise.resolve({});
    let validStatsVarPromise = Promise.resolve(new Set<string>());
    if (this.params.placeDcids.length !== 0) {
      placesPromise = getPlaceNames(this.params.placeDcids);
      validStatsVarPromise = getStatsVar(this.params.placeDcids);
    }
    Promise.all([
      statsVarInfoPromise,
      placesPromise,
      validStatsVarPromise,
    ]).then(([statsVarInfo, placeIdNames, statsVarValid]) => {
      this.setState({
        statsVarInfo: statsVarInfo,
        placeIdNames: placeIdNames,
        statsVarValid: statsVarValid,
      });
    });
  }

  // add one statsVar with nodePath
  private addStatsVar(statsVar: string, nodePath: string[]): void {
    if (this.params.addStatsVar(statsVar, nodePath)) {
      getStatsVarInfo(this.params.getStatsVarDcids()).then((data) => {
        this.setState({
          statsVarInfo: data,
          statsVarNodes: _.cloneDeep(this.params.statsVarNodes),
        });
      });
      this.params.setUrlStatsVars();
    }
  }

  // remove one statsVar with nodePath
  private removeStatsVar(statsVar: string, nodePath: string[] = []): void {
    if (this.params.removeStatsVar(statsVar, nodePath)) {
      const tempStatsVarInfo = this.state.statsVarInfo;
      if (!(statsVar in this.params.statsVarNodes)) {
        delete tempStatsVarInfo[statsVar];
      }
      this.setState({
        statsVarNodes: _.cloneDeep(this.params.statsVarNodes),
        statsVarInfo: tempStatsVarInfo,
      });
      this.params.setUrlStatsVars();
    }
  }

  // add one place
  private addPlace(place: string): void {
    if (this.params.addPlace(place)) {
      const placesPromise = getPlaceNames(this.params.placeDcids);
      const validStatsVarPromise = getStatsVar(this.params.placeDcids);
      Promise.all([placesPromise, validStatsVarPromise]).then((values) => {
        this.setState({
          placeIdNames: values[0],
          statsVarValid: values[1],
        });
      });
      this.params.setUrlPlaces();
    }
  }

  // remove one place
  private removePlace(place: string): void {
    if (this.params.removePLace(place)) {
      const tempPlace = this.state.placeIdNames;
      delete tempPlace[place];
      getStatsVar(this.params.placeDcids).then((data) => {
        this.setState({
          placeIdNames: tempPlace,
          statsVarValid: data,
        });
      });
      this.params.setUrlPlaces();
    }
  }

  // call back function passed down to menu for getting statsVar titles
  setStatsVarTitle(statsVarId2Title: Record<string, string>): void {
    for (const id in statsVarId2Title) {
      // remove title of unselected statsVars
      if (!Object.keys(this.state.statsVarNodes).includes(id)) {
        delete statsVarId2Title[id];
      }
    }
    this.setState({
      statsVarTitle: statsVarId2Title,
    });
  }

  // set PerCapita for a chart
  setChartPerCapita(mprop: string, pc: boolean): void {
    if (this.params.setChartPC(mprop, pc)) {
      this.setState({
        chartOptions: _.cloneDeep(this.params.chartOptions),
      });
      this.params.setUrlChartOptions();
    }
  }

  render(): JSX.Element {
    let statsVarFilter: StatsVarFilterInterface;
    if (Object.keys(this.state.placeIdNames).length === 0) {
      statsVarFilter = new NoopStatsVarFilter();
    } else {
      statsVarFilter = new TimelineStatsVarFilter(this.state.statsVarValid);
    }
    return (
      <div>
        <div className="explore-menu-container" id="explore">
          <div id="drill-scroll-container">
            <div className="title">Select variables:</div>
            <Menu
              selectedNodes={this.state.statsVarNodes}
              statsVarFilter={statsVarFilter}
              setStatsVarTitle={this.setStatsVarTitle.bind(this)}
              addStatsVar={this.addStatsVar.bind(this)}
              removeStatsVar={this.removeStatsVar.bind(this)}
            ></Menu>
          </div>
        </div>
        <div id="plot-container">
          <div className="container">
            <div id="search">
              <SearchBar
                places={this.state.placeIdNames}
                addPlace={this.addPlace.bind(this)}
                removePlace={this.removePlace.bind(this)}
              />
            </div>
            {Object.keys(this.state.placeIdNames).length === 0 && <Info />}
            {Object.keys(this.state.placeIdNames).length !== 0 &&
              Object.keys(this.state.statsVarInfo).length !== 0 && (
                <div id="chart-region">
                  <ChartRegion
                    places={this.state.placeIdNames}
                    statsVars={this.state.statsVarInfo}
                    statsVarTitle={this.state.statsVarTitle}
                    removeStatsVar={this.removeStatsVar.bind(this)}
                    chartOptions={this.state.chartOptions}
                    setPC={this.setChartPerCapita.bind(this)}
                  ></ChartRegion>
                </div>
              )}
          </div>
        </div>
      </div>
    );
  }
}
export { Page };
