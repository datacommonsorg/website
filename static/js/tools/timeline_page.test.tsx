/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

jest.mock("axios");
jest.mock("../chart/draw");
import Enzyme, { mount } from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import React from "react";
import axios from "axios";
import pretty from "pretty";
import { Page } from "./timeline_page";
import { axios_mock, drawGroupLineChart_mock } from "./mock_functions";

Enzyme.configure({ adapter: new Adapter() });

const globalAny: any = global;

test("Single place and single stats var", () => {
  globalAny.window = Object.create(window);
  document.body.innerHTML = '<button id="download-link"></div>';
  // Set url hash
  Object.defineProperty(window, "location", {
    writable: true,
    value: {
      hash: "#&place=geoId/05&statsVar=Median_Age_Person",
    },
  });
  // Mock drawGroupLineChart() as getComputedTextLength can has issue with jest
  // and jsdom.
  drawGroupLineChart_mock();
  // mock all the async axios call
  axios_mock();
  // Do the actual render!
  const wrapper = mount(<Page />);
  // There are 3 promises to resolve:
  // 1) all for [statsVarInfo, placeName, validStatsVar]
  // 2) get hierachy.json
  // 3) get stats
  return Promise.resolve(wrapper)
    .then(() => wrapper.update())
    .then(() => wrapper.update())
    .then(() => wrapper.update())
    .then(() => wrapper.update())
    .then(() => {
      wrapper.update();
      expect(wrapper.find("#chart-region").getDOMNode().innerHTML).toEqual(
        `<div class="card"><span class="chartPerCapita">Per capita<button class="perCapitaCheckbox">` +
          `</button><a href="/faq#perCapita"><span> *</span></a></span>` +
          `<div class="chart-svg"></div><div class="statsVarChipRegion">` +
          `<div class="pv-chip mdl-chip--deletable">` +
          `<span class="mdl-chip__text">Median age</span>` +
          `<button class="mdl-chip__action"><i class="material-icons">` +
          `cancel</i></button></div></div></div>`
      );

      // add one statsVar by clicking the checkbox
      wrapper.find("#drill .checkbox").at(0).simulate("click");
      Promise.resolve(wrapper)
        .then(() => wrapper.update())
        .then(() => {
          wrapper.update();
          expect(window.location.hash).toEqual(
            "place=geoId%2F05&statsVar=Median_Age_Person%2C0%2C2__Count_Person%2C0%2C0"
          );
          // test if the statsVar info and statsVarTitle is updated
          expect(wrapper.state().statsVarInfo).toEqual({
            Median_Age_Person: { md: "", mprop: "age", pt: "Person", pvs: {} },
            Count_Person: { md: "", mprop: "count", pt: "Person", pvs: {} },
          });
          expect(wrapper.state().statsVarTitle).toEqual({
            Median_Age_Person: "Median age",
            Count_Person: "Population",
          });
          expect(
            pretty(wrapper.find("#chart-region").getDOMNode().innerHTML)
          ).toMatchSnapshot();

          // delete one statsVar from the statsVar chips
          wrapper.find("#chart-region button i").at(1).simulate("click");
          Promise.resolve(wrapper)
            .then(() => wrapper.update())
            .then(() => {
              wrapper.update();
              expect(window.location.hash).toEqual(
                "place=geoId%2F05&statsVar=Median_Age_Person%2C0%2C2"
              );
              expect(wrapper.state().statsVarInfo).toEqual({
                Median_Age_Person: {
                  md: "",
                  mprop: "age",
                  pt: "Person",
                  pvs: {},
                },
              });
              expect(wrapper.state().statsVarTitle).toEqual({
                Median_Age_Person: "Median age",
              });
              expect(
                pretty(wrapper.find("#chart-region").getDOMNode().innerHTML)
              ).toMatchSnapshot();
              expect(
                pretty(wrapper.find("#drill").getDOMNode().innerHTML)
              ).toMatchSnapshot();
              // TODO(Lijuan): delete last statsVar by clicking checkbox
            });
        });
    });
});

// // TODO(Lijuan): two places and one statsVar, statsVar valid for one, invalid for the other

test("statsVar not in PV-tree", () => {
  Object.defineProperty(window, "location", {
    value: {
      hash: "#&place=geoId/05&statsVar=NotInTheTree",
    },
  });
  // Mock drawGroupLineChart() as getComputedTextLength can has issue with jest
  // and jsdom.
  drawGroupLineChart_mock();
  // mock all the async axios call
  axios_mock();
  // Do the actual render!
  const wrapper = mount(<Page />);
  // There are 3 promises to resolve:
  // 1) all for [statsVarInfo, placeName, validStatsVar]
  // 2) get hierachy.json
  // 3) get stats
  return Promise.resolve(wrapper)
    .then(() => wrapper.update())
    .then(() => wrapper.update())
    .then(() => wrapper.update())
    .then(() => wrapper.update())
    .then(() => {
      wrapper.update();
      expect(
        pretty(wrapper.find("#chart-region").getDOMNode().innerHTML)
      ).toMatchSnapshot();
      expect(
        pretty(wrapper.find("#drill").getDOMNode().innerHTML)
      ).toMatchSnapshot();
    });
});

test("chart options", () => {
  // test if the chart options are correctly set
  Object.defineProperty(window, "location", {
    writable: true,
    value: {
      hash: "#&place=geoId/05&statsVar=Median_Age_Person",
    },
  });

  // Mock drawGroupLineChart() as getComputedTextLength can has issue with jest
  // and jsdom.
  drawGroupLineChart_mock();
  // mock all the async axios call
  axios_mock();

  // Do the actual render!
  const wrapper = mount(<Page />);
  // There are 3 promises to resolve:
  // 1) all for [statsVarInfo, placeName, validStatsVar]
  // 2) get hierachy.json
  // 3) get stats
  return Promise.resolve(wrapper)
    .then(() => wrapper.update())
    .then(() => wrapper.update())
    .then(() => wrapper.update())
    .then(() => wrapper.update())
    .then(() => {
      wrapper.update();
      // set per capita to True
      wrapper.find("#chart-region .perCapitaCheckbox").simulate("click");
      Promise.resolve(wrapper)
        .then(() => wrapper.update())
        .then(() => {
          wrapper.update();
          expect(window.location.hash).toBe(
            "place=geoId%2F05&statsVar=Median_Age_Person&chart=%7B%22age%22%3A%7B%22pc%22%3Atrue%7D%7D"
          );
          expect(
            pretty(wrapper.find("#chart-region").getDOMNode().innerHTML)
          ).toMatchSnapshot();

          // remove the statsVar
          wrapper.find("#drill .checkbox.checked").simulate("click");
          expect(wrapper.find("#chart-region").length).toBe(0); // chart deleted
          expect(window.location.hash).toBe(
            "place=geoId%2F05&statsVar=&chart=%7B%22age%22%3A%7B%22pc%22%3Afalse%7D%7D"
          );

          // TODO: look into the mock functions
          // Note: when the same url is called the second time, the returned value is {}
          axios_mock();
          // add the same statsVar back
          wrapper.find("#drill .checkbox").at(1).simulate("click");
          expect(axios.get).toHaveBeenLastCalledWith(
            "/api/stats/stats-var-property?dcid=Median_Age_Person"
          );
          expect(window.location.hash).toBe(
            "place=geoId%2F05&statsVar=Median_Age_Person%2C0%2C2&chart=%7B%22age%22%3A%7B%22pc%22%3Afalse%7D%7D"
          );

          Promise.resolve(wrapper)
            .then(() => wrapper.update())
            .then(() => {
              wrapper.update();
              expect(
                pretty(wrapper.find("#chart-region").getDOMNode().innerHTML)
              ).toMatchSnapshot();
            });
        });
    });
});
