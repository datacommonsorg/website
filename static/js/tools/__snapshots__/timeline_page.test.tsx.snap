// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Single place and single stats var 1`] = `
"<div class=\\"card\\"><span class=\\"chartPerCapita\\">Per capita<button class=\\"perCapitaCheckbox\\"></button></span>
  <div class=\\"chart-svg\\"></div>
  <div class=\\"statsVarChipRegion\\">
    <div class=\\"pv-chip mdl-chip--deletable\\"><span class=\\"mdl-chip__text\\">Median age</span><button class=\\"mdl-chip__action\\"><i class=\\"material-icons\\">cancel</i></button></div>
  </div>
</div>
<div class=\\"card\\"><span class=\\"chartPerCapita\\">Per capita<button class=\\"perCapitaCheckbox\\"></button></span>
  <div class=\\"chart-svg\\"></div>
  <div class=\\"statsVarChipRegion\\">
    <div class=\\"pv-chip mdl-chip--deletable\\"><span class=\\"mdl-chip__text\\">Population</span><button class=\\"mdl-chip__action\\"><i class=\\"material-icons\\">cancel</i></button></div>
  </div>
</div>"
`;

exports[`Single place and single stats var 2`] = `
"<div class=\\"card\\"><span class=\\"chartPerCapita\\">Per capita<button class=\\"perCapitaCheckbox\\"></button></span>
  <div class=\\"chart-svg\\"><svg width=\\"500\\" height=\\"500\\"><text>svg text</text></svg></div>
  <div class=\\"statsVarChipRegion\\">
    <div class=\\"pv-chip mdl-chip--deletable\\"><span class=\\"mdl-chip__text\\">Median age</span><button class=\\"mdl-chip__action\\"><i class=\\"material-icons\\">cancel</i></button></div>
  </div>
</div>"
`;

exports[`Single place and single stats var 3`] = `
"<div class=\\"noedge\\">
  <ul class=\\"noborder\\">
    <li class=\\"value\\" id=\\"Demographics\\"><span><a class=\\"value-link\\">Demographics  <sup>(1787)</sup><img class=\\"right-caret transform-up\\" src=\\"/images/right-caret-light.png\\"></a></span>
      <ul class=\\"noborder\\">
        <li class=\\"value\\" id=\\"Population\\"><span><a class=\\"value-link\\">Population  <sup>(1)</sup><button class=\\"checkbox\\"></button></a></span></li>
      </ul>
      <ul class=\\"noborder\\">
        <li class=\\"value\\" id=\\"Median age\\"><span><a class=\\"value-link\\">Median age  <sup>(1)</sup><button class=\\"checkbox checked\\"></button></a></span></li>
      </ul>
    </li>
  </ul>
</div>"
`;
